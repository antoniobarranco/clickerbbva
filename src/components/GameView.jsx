import React, { useEffect, useState } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';

import { useBeforeunload } from 'react-beforeunload';
import {
  GameValues, BaseGameState, calcClickersIncrement, calcBoostersCost, formatNumericValue,
} from '../helpers/GameLogic';
import {
  CheckUserExists, CreateUser, LoadUser, SaveUser,
} from '../helpers/UserLogic';
import GameViewScore from './GameViewScore';
import GameViewControls from './GameViewControls';
import GameViewStats from './GameViewStats';
import BreadCrumbs from './ui/BreadCrumbs';

function GameView() {
  //
  // This is the game view of the app. Because all game logic resides on that view and this is
  // relatively simple, all states, and side effects are maintaned on this component, following
  // Flux pattern instead a less simple Redux or React Context approach which justifies on more
  // complex applications and reduces component independence by sharing external or non-facilitated
  // state through its props
  //

  //
  // States Set: there are a main game state based on player data and four simple states to
  // control the activation or booster buttons
  //

  let gameInitialState = BaseGameState;
  const [gameState, setGameState] = useState(gameInitialState);
  const [canBuyAutoClicker, setcanBuyAutoClicker] = useState(false);
  const [canBuyMegaClicker, setcanBuyMegaClicker] = useState(false);
  const [autoClickerCost, setautoClickerCost] = useState(0);
  const [megaClickerCost, setmegaClickerCost] = useState(0);

  //
  // Get de params of URL sanityzed bi queryString
  //

  const navigate = useNavigate();
  const location = useLocation();

  const { player } = queryString.parse(location.search);

  //
  // Get default inital player state and get or set depending of user match on Local Storage
  //

  useEffect(() => {
    if (!player) return navigate('/', { replace: true });

    if (CheckUserExists(player)) {
      gameInitialState = { ...LoadUser(player) };
    } else {
      gameInitialState = { ...CreateUser(player) };
    }

    setGameState(gameInitialState);
  }, [player]);

  //
  // Handles the exit button in order to save current state.
  // Also handles the hard quit using a hook that listen BeforeUnload event
  //

  const handleExit = () => {
    SaveUser(gameState, player);
    navigate('/', { replace: true });
  };

  useBeforeunload(handleExit);

  //
  // Handles button interaction. They work updating the proper state
  //

  const hitHandle = () => {
    const scoreIncrement = gameState.score + GameValues.normalIncrement;
    setGameState({ ...gameState, score: scoreIncrement });
  };

  const autoClickerHandle = () => {
    const autoClickersIncrement = gameState.autoclickers + 1;
    setGameState({ ...gameState, autoclickers: autoClickersIncrement });
  };

  const megaClickerHandle = () => {
    const megaClickersIncrement = gameState.megaclickers + 1;
    setGameState({ ...gameState, megaclickers: megaClickersIncrement });
  };

  //
  // To update UI states that are used to show or hide booster buttons
  //

  const updateBoosterAvailability = () => {
    const { score } = gameState;
    setcanBuyAutoClicker((score >= autoClickerCost));
    setcanBuyMegaClicker((score >= megaClickerCost));
    // (score >= autoClickerCost) ? setcanBuyAutoClicker(true) : setcanBuyAutoClicker(false);
    // (score >= megaClickerCost) ? setcanBuyMegaClicker(true) : setcanBuyMegaClicker(false);
  };

  //
  // Game loop: uses and useEffect to create a side effect that avoids counting errors because
  // async nature of React renders. Game loop updates get pre calculed values and updates in
  // one time operation, avoiding the use of more side effects or even more timers.
  //

  useEffect(() => {
    setTimeout(() => {
      const BoosterCost = calcBoostersCost(gameState.autoclickers, gameState.megaclickers);

      setautoClickerCost(BoosterCost.autoClickerCost);
      setmegaClickerCost(BoosterCost.megaClickerCost);
      updateBoosterAvailability(BoosterCost);

      const { autoclickers, megaclickers } = gameState;

      const clickersIncrement = calcClickersIncrement(autoclickers, megaclickers);

      setGameState((gameState) => ({ ...gameState, score: gameState.score + clickersIncrement }));
    }, GameValues.timeFrequency);
  });

  return (
    <>

      <BreadCrumbs Text={`Hi ${gameState.user}!`} handleExit={handleExit} />

      <GameViewScore score={formatNumericValue(gameState.score)} />

      <GameViewControls
        hitHandle={hitHandle}
        autoClickerHandle={autoClickerHandle}
        megaClickerHandle={megaClickerHandle}
        autoClickers={gameState.autoclickers}
        megaClickers={gameState.megaclickers}
        canBuyAutoClicker={canBuyAutoClicker}
        canBuyMegaClicker={canBuyMegaClicker}
      />

      <GameViewStats
        clickersIncrement={calcClickersIncrement(gameState.autoclickers, gameState.megaclickers)}
        autoClickerCost={autoClickerCost}
        megaClickerCost={megaClickerCost}
        timeInterval={GameValues.timeFrequency}
      />

    </>
  );
}

export default GameView;
