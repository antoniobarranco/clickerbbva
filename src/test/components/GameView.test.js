import renderer from 'react-test-renderer';
import GameView from "../../components/GameView";

import { MemoryRouter } from "react-router-dom";

describe('Tests on GameView component', () => {

    test('Test GameView with Snapshot test for UI changes', () => {

      const tree = renderer.create(
        <MemoryRouter initialEntries={['/dashboard']}>
          <GameView />
        </MemoryRouter>,).toJSON();

      expect(tree).toMatchSnapshot();

    })

})