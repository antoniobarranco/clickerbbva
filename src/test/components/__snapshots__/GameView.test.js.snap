// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on GameView component Test GameView with Snapshot test for UI changes 1`] = `
Array [
  <nav
    className="navbar navbar-dark bg-primary bg-opacity-75"
  >
    <div
      className="container-fluid"
    >
      <span
        className="navbar-brand mb-0 h1"
        data-testid="BreadCrumbsText"
      >
        Hi unnamed!
      </span>
      <div
        className="d-flex"
      >
        <button
          className="btn btn-outline-light"
          onClick={[Function]}
          type="button"
        >
          <i
            className="bi bi-box-arrow-in-right"
          />
           
          Close
        </button>
      </div>
    </div>
  </nav>,
  <div
    className="d-grid gap-2 col-6 my-5 mx-auto"
  >
    <h1
      className="display-1 text-center"
    >
      <pre>
        <span
          className="badge bg-secondary"
        >
          0 Hits
        </span>
      </pre>
    </h1>
  </div>,
  <div
    className="d-grid gap-2 col-6  my-5 mx-auto"
  >
    <button
      className="btn btn-primary btn-lg"
      data-testid="HitButton"
      onClick={[Function]}
      type="button"
    >
      Hit
    </button>
    <button
      aria-disabled={true}
      className="btn btn-warning btn-lg"
      data-testid="autoClickerButton"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <i
        className="bi bi-star"
      />
       
      Buy AutoClicker
       
      <span
        className="badge text-bg-light"
      >
        x
        0
      </span>
    </button>
    <button
      aria-disabled={true}
      className="btn btn-danger btn-lg"
      data-testid="megaClickerButton"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <i
        className="bi bi-star-fill"
      />
       
      Buy MegaClicker
       
      <span
        className="badge text-bg-light"
      >
        x
        0
      </span>
    </button>
  </div>,
  <div
    className="d-grid gap-2 col-8  my-5 mx-auto"
  >
    <ul
      className="list-group"
    >
      <li
        className="list-group-item d-flex justify-content-between align-items-center active"
      >
        Game Stats
      </li>
      <li
        className="list-group-item d-flex justify-content-between align-items-center"
      >
        Increment every
         
        100
        ms
        <span
          className="badge bg-primary rounded-pill"
          data-testid="clickersIncrement"
        >
          0
        </span>
      </li>
      <li
        className="list-group-item d-flex justify-content-between align-items-center"
      >
        AutoClickers cost
         
        <span
          className="badge bg-primary rounded-pill"
          data-testid="autoClickerCost"
        >
          0
        </span>
      </li>
      <li
        className="list-group-item d-flex justify-content-between align-items-center"
      >
        MegaClickers cost
         
        <span
          className="badge bg-primary rounded-pill"
          data-testid="megaClickerCost"
        >
          0
        </span>
      </li>
    </ul>
  </div>,
]
`;
